### PR TITLE
xla: add aarch64-darwin support

### LINKED ISSUE #390395: bazel_7: aarch64-darwin build broken with enableNixHacks, necessary for buildBazelPackage
### Nixpkgs version

- Unstable (25.05)

### Describe the bug

On 9ea8b8558b06d3f26ae135078efa534f319db478,
```
❯ nix-build -E 'with import <nixpkgs> {}; bazel_7.override { enableNixHacks = true; }'
...
ERROR: no such package '@@bazel_tools~xcode_configure_extension~local_config_xcode//': BUILD file not found in directory '' of external repository @@bazel_tools~xcode_configure_extension~local_config_xcode. Add a BUILD file to a directory to mark it as a package.
ERROR: /private/tmp/nix-build-bazel-7.4.1.drv-0/bazel_bwZHPRsM/out/external/bazel_tools/tools/cpp/BUILD:125:6: no such package '@@bazel_tools~xcode_configure_extension~local_config_xcode//': BUILD file not found in directory '' of external repository @@bazel_tools~xcode_configure_extension~local_config_xcode. Add a BUILD file to a directory to mark it as a package. and referenced by '@@bazel_tools//tools/cpp:host_xcodes'
ERROR: Analysis of target '//src:bazel_nojdk' failed; build aborted: Analysis failed
INFO: Elapsed time: 5.087s, Critical Path: 0.02s
INFO: 1 process: 1 internal.
ERROR: Build did NOT complete successfully
FAILED: 

ERROR: Could not build Bazel
```

Full log: https://gist.github.com/samuela/b0eadb6fc95508f51efd0aaeb4b3e590

Note that this build is necessary for a version of bazel to be usable by `buildBazelPackage` due to https://github.com/NixOS/nixpkgs/blob/f24a24b224695efc08ab1d1091e8c30bae55b441/pkgs/build-support/build-bazel-package/default.nix#L213

### Steps to reproduce

`nix-build -E 'with import <nixpkgs> {}; bazel_7.override { enableNixHacks = true; }'`

### Expected behaviour

successful build

### Screenshots

_No response_

### Relevant log output

```console

```

### Additional context

_No response_

### System metadata

 - system: `"aarch64-darwin"`
 - host os: `Darwin 24.3.0, macOS 15.3.2`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.24.9`
 - channels(samuelainsworth): `"darwin, home-manager, nixpkgs-23.11pre511546.844ffa82bbe2"`
 - channels(root): `"nixpkgs"`
 - nixpkgs: `/Users/samuelainsworth/.nix-defexpr/channels/nixpkgs`

### Notify maintainers

@mboes @uri-canva @cbley @olebedev @groodt @aherrmann @ylecornec

---

**Note for maintainers:** Please tag this issue in your pull request description. (i.e. `Resolves #ISSUE`.)

### I assert that this issue is relevant for Nixpkgs

- [x] I assert that this is a bug and not a support request.
- [x] I assert that this is not a [duplicate of an existing issue](https://github.com/NixOS/nixpkgs/issues?q=is%3Aopen+is%3Aissue+label%3A%220.kind%3A+bug%22+-label%3A%226.topic%3A+darwin%22+-label%3A%226.topic%3A+nixos%22). 
- [x] I assert that I have read the [NixOS Code of Conduct](https://github.com/NixOS/.github/blob/master/CODE_OF_CONDUCT.md) and agree to abide by it.

### Is this issue important to you?

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

### PR DESCRIPTION
**Pull request by samuela**
## Summary

- **bazel_7**: Fix `enableNixHacks` patch to check for `BUILD`/`BUILD.bazel`/`WORKSPACE`/`WORKSPACE.bazel` before treating cached repos as up-to-date. The previous unconditional skip caused darwin builds to fail because empty dirs created by module extensions were returned before repository rules could run.
- **xla**: Add aarch64-darwin (Apple Silicon) support:
  - Add `cctools.libtool` for darwin toolchain detection
  - Set macOS deployment target (`-mmacosx-version-min=11.0`) via `--copt`/`--host_copt`/`--linkopt`/`--host_linkopt` to fix `futimens` availability errors in LLVM host tool builds
  - Add per-platform `fetchAttrs.sha256` (preserves existing x86_64-linux hash)
  - Remove `badPlatforms` since darwin now builds successfully

Fixes https://github.com/NixOS/nixpkgs/issues/390395

### CODE REVIEW

**Comment by GaetanLepage** [comment]
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 492814`
Commit: `d73a3b2875aff836d2788121566a9119ef7a0b86`

---
### `x86_64-linux`
<details>
  <summary>:x: 3 packages failed to build:</summary>
  <ul>
    <li>envoy</li>
    <li>perf_data_converter</li>
    <li>pomerium</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 33 packages built:</summary>
  <ul>
    <li>anytype</li>
    <li>anytype-heart</li>
    <li>bant</li>
    <li>fcitx5-mozc</li>
    <li>fcitx5-mozc-ut</li>
    <li>mozc (ibus-engines.mozc)</li>
    <li>mozc-ut (ibus-engines.mozc-ut)</li>
    <li>protoc-gen-js</li>
    <li>python313Packages.arviz</li>
    <li>python313Packages.arviz.dist</li>
    <li>python313Packages.bambi</li>
    <li>python313Packages.bambi.dist</li>
    <li>python313Packages.corner</li>
    <li>python313Packages.corner.dist</li>
    <li>python313Packages.distrax</li>
    <li>python313Packages.distrax.dist</li>
    <li>python313Packages.flowjax</li>
    <li>python313Packages.flowjax.dist</li>
    <li>python313Packages.numpyro</li>
    <li>python313Packages.numpyro.dist</li>
    <li>python313Packages.nutpie</li>
    <li>python313Packages.nutpie.dist</li>
    <li>python313Packages.pymc</li>
    <li>python313Packages.pymc.dist</li>
    <li>python313Packages.pytensor</li>
    <li>python313Packages.pytensor.dist</li>
    <li>python313Packages.rlax</li>
    <li>python313Packages.rlax.dist</li>
    <li>python313Packages.tensorflow-probability</li>
    <li>python313Packages.tensorflow-probability.dist</li>
    <li>verible</li>
    <li>veridian</li>
    <li>xla</li>
  </ul>
</details>

---
### `aarch64-linux`
<details>
  <summary>:x: 1 package failed to build:</summary>
  <ul>
    <li>perf_data_converter</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 34 packages built:</summary>
  <ul>
    <li>anytype</li>
    <li>anytype-heart</li>
    <li>bant</li>
    <li>envoy</li>
    <li>fcitx5-mozc</li>
    <li>fcitx5-mozc-ut</li>
    <li>mozc (ibus-engines.mozc)</li>
    <li>mozc-ut (ibus-engines.mozc-ut)</li>
    <li>pomerium</li>
    <li>protoc-gen-js</li>
    <li>python313Packages.arviz</li>
    <li>python313Packages.arviz.dist</li>
    <li>python313Packages.bambi</li>
    <li>python313Packages.bambi.dist</li>
    <li>python313Packages.corner</li>
    <li>python313Packages.corner.dist</li>
    <li>python313Packages.distrax</li>
    <li>python313Packages.distrax.dist</li>
    <li>python313Packages.flowjax</li>
    <li>python313Packages.flowjax.dist</li>
    <li>python313Packages.numpyro</li>
    <li>python313Packages.numpyro.dist</li>
    <li>python313Packages.nutpie</li>
    <li>python313Packages.nutpie.dist</li>
    <li>python313Packages.pymc</li>
    <li>python313Packages.pymc.dist</li>
    <li>python313Packages.pytensor</li>
    <li>python313Packages.pytensor.dist</li>
    <li>python313Packages.rlax</li>
    <li>python313Packages.rlax.dist</li>
    <li>python313Packages.tensorflow-probability</li>
    <li>python313Packages.tensorflow-probability.dist</li>
    <li>verible</li>
    <li>veridian</li>
  </ul>
</details>

---
### `x86_64-darwin`
<details>
  <summary>:x: 1 package failed to build:</summary>
  <ul>
    <li>protoc-gen-js</li>
  </ul>
</details>

---
### `aarch64-darwin`
<details>
  <summary>:x: 3 packages failed to build:</summary>
  <ul>
    <li>protoc-gen-js</li>
    <li>verible</li>
    <li>xla</li>
  </ul>
</details>

---
<details>
<summary>Error logs: `x86_64-linux`</summary>
<details>
<summary>perf_data_converter</summary>
<pre>                 from external/abseil-cpp~/absl/flags/reflection.h:29:
external/abseil-cpp~/absl/container/internal/container_memory.h: In function &#x27;void* absl::lts_20240116::container_internal::Allocate(Alloc*, size_t)&#x27;:
external/abseil-cpp~/absl/container/internal/container_memory.h:66:27: error: &#x27;uintptr_t&#x27; does not name a type [-Wtemplate-body]
   66 |   assert(reinterpret_cast&lt;uintptr_t&gt;(p) % Alignment == 0 &amp;&amp;
      |                           ^~~~~~~~~
external/abseil-cpp~/absl/container/internal/container_memory.h:31:1: note: &#x27;uintptr_t&#x27; is defined in header &#x27;&lt;cstdint&gt;&#x27;; this is probably fixable by adding &#x27;#include &lt;cstdint&gt;&#x27;
   30 | #include &quot;absl/utility/utility.h&quot;
  +++ |+#include &lt;cstdint&gt;
   31 | 
Use --verbose_failures to see the command lines of failed build steps.
INFO: Elapsed time: 4.917s, Critical Path: 2.54s
INFO: 838 processes: 599 internal, 239 linux-sandbox.
ERROR: Build did NOT complete successfully
//src:intervalmap_test                                                NO STATUS
//src:perf_data_converter_test                                        NO STATUS
//src:perf_to_profile_lib_test                                        NO STATUS
//src:perf_data_handler_test                                    FAILED TO BUILD

Executed 0 out of 4 tests: 1 fails to build and 3 were skipped.
</pre>
</details>
</details>

---
<details>
<summary>Error logs: `aarch64-linux`</summary>
<details>
<summary>perf_data_converter</summary>
<pre>                 from external/abseil-cpp~/absl/flags/reflection.h:29:
external/abseil-cpp~/absl/container/internal/container_memory.h: In function &#x27;void* absl::lts_20240116::container_internal::Allocate(Alloc*, size_t)&#x27;:
external/abseil-cpp~/absl/container/internal/container_memory.h:66:27: error: &#x27;uintptr_t&#x27; does not name a type [-Wtemplate-body]
   66 |   assert(reinterpret_cast&lt;uintptr_t&gt;(p) % Alignment == 0 &amp;&amp;
      |                           ^~~~~~~~~
external/abseil-cpp~/absl/container/internal/container_memory.h:31:1: note: &#x27;uintptr_t&#x27; is defined in header &#x27;&lt;cstdint&gt;&#x27;; this is probably fixable by adding &#x27;#include &lt;cstdint&gt;&#x27;
   30 | #include &quot;absl/utility/utility.h&quot;
  +++ |+#include &lt;cstdint&gt;
   31 | 
Use --verbose_failures to see the command lines of failed build steps.
INFO: Elapsed time: 32.363s, Critical Path: 11.33s
INFO: 1572 processes: 646 internal, 926 linux-sandbox.
ERROR: Build did NOT complete successfully
//src:perf_data_converter_test                                        NO STATUS
//src:perf_to_profile_lib_test                                        NO STATUS
//src:perf_data_handler_test                                    FAILED TO BUILD
//src:intervalmap_test                                                   PASSED in 0.0s

Executed 1 out of 4 tests: 1 test passes, 1 fails to build and 2 were skipped.
</pre>
</details>
</details>

---
<details>
<summary>Error logs: `x86_64-darwin`</summary>
<details>
<summary>protoc-gen-js</summary>
<pre>In file included from external/com_google_absl/absl/base/nullability.h:153:
In file included from external/com_google_absl/absl/base/internal/nullability_impl.h:22:
external/com_google_absl/absl/meta/type_traits.h:511:36: warning: builtin __is_trivially_relocatable is deprecated; use __builtin_is_cpp_trivially_relocatable instead [-Wdeprecated-builtins]
  511 |     : std::integral_constant&lt;bool, __is_trivially_relocatable(T)&gt; {};
      |                                    ^
In file included from external/com_google_absl/absl/strings/escaping.cc:15:
In file included from external/com_google_absl/absl/strings/escaping.h:32:
In file included from external/com_google_absl/absl/strings/str_join.h:59:
In file included from external/com_google_absl/absl/strings/internal/str_join_internal.h:43:
In file included from external/com_google_absl/absl/strings/str_cat.h:108:
external/com_google_absl/absl/strings/numbers.h:27:10: error: module com_google_absl//absl/strings:strings does not depend on a module exporting &#x27;tmmintrin.h&#x27;
   27 | #include &lt;tmmintrin.h&gt;
      |          ^
1 warning and 1 error generated.
Target //generator:protoc-gen-js failed to build
Use --verbose_failures to see the command lines of failed build steps.
INFO: Elapsed time: 40.244s, Critical Path: 1.62s
INFO: 1158 processes: 1027 internal, 131 processwrapper-sandbox.
ERROR: Build did NOT complete successfully
</pre>
</details>
</details>

---
<details>
<summary>Error logs: `aarch64-darwin`</summary>
<details>
<summary>protoc-gen-js</summary>
<pre>ERROR: /nix/var/nix/builds/nix-58768-3729406321/output/external/com_google_protobuf/third_party/utf8_range/BUILD.bazel:32:11: Compiling third_party/utf8_range/utf8_range.c failed: (Exit 1): cc_wrapper.sh failed: error executing CppCompile command (from target @@com_google_protobuf//third_party/utf8_range:utf8_range) external/local_config_cc/cc_wrapper.sh -U_FORTIFY_SOURCE -fstack-protector -Wall -Wthread-safety -Wself-assign -Wunused-but-set-parameter -Wno-free-nonheap-object -fcolor-diagnostics ... (remaining 61 arguments skipped)

Use --sandbox_debug to see verbose messages from the sandbox and retain the sandbox build root for debugging
In file included from external/com_google_protobuf/third_party/utf8_range/utf8_range.c:18:
external/com_google_protobuf/third_party/utf8_range/utf8_range.h:4:10: error: module com_google_protobuf//third_party/utf8_range:utf8_range does not depend on a module exporting &#x27;stddef.h&#x27;
    4 | #include &lt;stddef.h&gt;
      |          ^
external/com_google_protobuf/third_party/utf8_range/utf8_range.c:20:10: error: module com_google_protobuf//third_party/utf8_range:utf8_range does not depend on a module exporting &#x27;stddef.h&#x27;
   20 | #include &lt;stddef.h&gt;
      |          ^
external/com_google_protobuf/third_party/utf8_range/utf8_range.c:21:10: error: module com_google_protobuf//third_party/utf8_range:utf8_range does not depend on a module exporting &#x27;stdint.h&#x27;
   21 | #include &lt;stdint.h&gt;
      |          ^
3 errors generated.
Target //generator:protoc-gen-js failed to build
Use --verbose_failures to see the command lines of failed build steps.
INFO: Elapsed time: 5.286s, Critical Path: 1.11s
INFO: 838 processes: 832 internal, 6 processwrapper-sandbox.
ERROR: Build did NOT complete successfully
</pre>
</details>
<details>
<summary>verible</summary>
<pre>//verible/verilog/tools/lint:verilog-syntax-lexer-fail_test           NO STATUS
//verible/verilog/tools/lint:void-cast-allrules_test                  NO STATUS
//verible/verilog/tools/lint:void-cast-defaultrules_test              NO STATUS
//verible/verilog/tools/lint:void-cast-minusrule_test                 NO STATUS
//verible/verilog/tools/lint:void-cast-onerule_test                   NO STATUS
//verible/verilog/tools/lint:void-cast-syntax_test                    NO STATUS
//verible/verilog/tools/ls:autoexpand_test                            NO STATUS
//verible/verilog/tools/ls:lsp-parse-buffer_test                      NO STATUS
//verible/verilog/tools/ls:symbol-table-handler_test                  NO STATUS
//verible/verilog/tools/ls:verible-verilog-ls_test                    NO STATUS
//verible/verilog/tools/ls:verilog-language-server_test               NO STATUS
//verible/verilog/tools/obfuscator:obfuscate_test                     NO STATUS
//verible/verilog/tools/preprocessor:verilog-preprocessor_test        NO STATUS
//verible/verilog/tools/project:project-tool_test                     NO STATUS
//verible/verilog/tools/syntax:verilog-syntax_test                    NO STATUS
//verible/verilog/transform:obfuscate_test                            NO STATUS
//verible/verilog/transform:strip-comments_test                       NO STATUS

Executed 0 out of 608 tests: 608 were skipped.
</pre>
</details>
</details>

**Comment by samuela** [comment]
I am finding the following differences:
  1. go_sdk: Yours has linux_amd64 packages, mine has darwin_arm64
  2. onednn: Present on Gaetan's, absent on mine
  3. local_config_cuda: Minor differences (.dylib vs .so stubs)

From the `go_sdk/BUILD.bazel` in your tarball:
```
  goarch = "amd64",
  goos = "linux",
  host = "linux_amd64",
```
which leads me to suspect that this build was not run on aarch64-darwin or somehow the builder is confused about what system it should be running on. Did you run these commands via emulation or similar? What type of physical hardware was used? I am tentatively guessing that some platform dependent stuff is sneaking in through nixpkgs-review.

**Comment by GaetanLepage** [comment]
Oh wait, I realize what happened. The failure is from `nixpkgs-review` which had run the darwin build on a remote darwin builder.
When I extracted the xla deps as per your command, I naively did it on my linux system. Sorry for the confusion.

Here is the result from an actual mac:
https://filesender.renater.fr/?s=download&token=977da611-5dee-49a7-8eee-09125d59692b

**Comment by GaetanLepage** [comment]
Nice! Seems like it works

**Comment by samuela** [comment]
Great, thanks for the assist @GaetanLepage ! I'll merge tomorrow unless anyone objects?

**Review by GaetanLepage — APPROVED** [review]
